frontend/test/metabase/lib/expressions/generator.js

```diff
@@ -16,7 +16,7 @@ const sfc32 = (a, b, c, d) => {
   };
 };
 
-export function generateExpression(seed) {
+export function generateExpression(seed, type) {
```

Comment:
The density of logic crammed into this one function makes it hard to follow and review the file. Can you break it up into a utils factory for all the functions that just close over `seed` and then a factory function to close over `depth` + `seed`? I think the following would be alot cleaner as the entry point at the cost of passing `seed` around some more:

```typescript

function generateTokenSeq(seed) {
  const expressions = expressionsFactory(seed);
  const tree = expressions.expression();
  return { tree, expression: format(tree) };
}

function generateValidSyntax(seed, type) {
  const expressions = expressionsFactory(seed);

  let tree;
  switch (type) {
    case "boolean":
    case "filter":
      tree = expressions.boolean();
      break;

    case "number":
      tree = expressions.number();
      break;

    case "string":
      tree = expressions.string();
      break;

    case "expression":
      tree = expressions.expression();
      break;

    default:
      throw new Error(`Unknown expression type ${type}`);
  }

  return { tree, expression: format(tree) };
}

function generateExpression(seed, type) {
  return type ? generateValidSyntax(seed, type) : generateTokenSeq(seed);
}
```

Reply:
I don't know how to make it work like that since there is a difference between `expressionFactory` for `generateTokenSeq` vs `generateValidSyntax`. But maybe that's where the complication emerged: because I still wanted to preserve the current way the valid tokens are generated (which likely not yield a syntactically valid expression, since the grammar isn't followed).

I've tweaked the code to focus on a _single_ way of producing expressions, and the expressions are guaranteed to be syntactically valid, and let's use the same generation method to fuzz both the tokenizer and the parser. Please take a look again and this variant should be much simpler now.